src/gnu/trove/array/AbstractOffheapArray.java

```diff
@@ -141,4 +149,68 @@ public Field run() {
             }
         });
     }
+
+    private static Object createCleaner(Object obj, Runnable cleanUpRunnable) {
+        if (createMethod == null) {
+            initializeCreateMethod();
+        }
+        try {
+            return createMethod.invoke(null, obj, cleanUpRunnable);
+        } catch (ReflectiveOperationException e) {
+            throw new IllegalStateException("Could not create a new cleaner.", e);
+        }
+    }
+
+    private static synchronized void initializeCreateMethod() {
+        if (createMethod != null) {
+            return;
+        }
+
+        try {
+            Class<?> cleaner = Class.forName("sun.misc.Cleaner");
+            createMethod = cleaner.getMethod("create", Object.class, Runnable.class);
+            return;
+        } catch (ReflectiveOperationException e) {
+            // means we are on java 9+, try different class
+        }
+
+        try {
+            Class<?> cleaner = Class.forName("jdk.internal.ref.Cleaner");
+            createMethod = cleaner.getMethod("create", Object.class, Runnable.class);
+        } catch (ReflectiveOperationException e) {
+            throw new IllegalStateException("Cannot find Cleaner to use.", e);
+        }
+    }
+
+    private void clean() {
+        if (cleanMethod == null) {
+            initializeCleanMethod();
+        }
+        try {
+            cleanMethod.invoke(cleaner);
+        } catch (ReflectiveOperationException e) {
+            throw new IllegalStateException("Could not invoke clean method.", e);
+        }
+    }
+
+    private static synchronized void initializeCleanMethod() {
+        if (cleanMethod != null) {
```

Comment:
Double check locking is a bit wonky here without the `volatile` keyword.

I don't think we really need volatile reads though, perhaps we could update these methods to return the `Method` object instead of mutating static state, that way we can use `private static final` values where these are called on class init.

Instead of failing, we'll need to return null, and add null checks to their accessors.

Reply:
Out of curiosity, what is the big downside of not doing that? Basically, I don't think it is that bad if this is called a few times. I'm happy to make the change since it is _more_ correct in some sense, but I just want to know if I am missing something big here or if it is just to be cautious.

Reply:
There's a good description on the error-prone check: https://errorprone.info/bugpattern/DoubleCheckedLocking